src/System/Terminal/Widgets/SearchSelect.hs

```diff
@@ -0,0 +1,126 @@
+-- selections to be searched and sorted via the user's text input, similar to
+-- [vertico for Emacs](https://github.com/minad/vertico).
+--
+-- Note: FIXME: A 'Map' is used internally. Therefore, the order of the search
+-- result is determined by the output of 'Data.Map.toList'. For as long as 'Select'
+-- continues to use a list, it may be preferable to use the converted output
+-- to initialize 'selectWidget'. This prevents the jarring transition upon
+-- first user input.
+--
+-- Example:
+-- >>> textInput = TextInput "M-x " False False mempty id
+-- >>> opts = Map.toList $ Map.fromList [("foo", False), ("bar", True), ("baz", False)]
+-- >>> select = Select "Options" opts False 0
+-- >>> searchSelect = SearchSelect textInput select (Map.fromList opts) True 10
+-- >>> withTerminal . runTerminalT $ runWidget searchSelect
+-- @
+-- M-x <search>
+-- Options
+--  (*) [bar]
+--  ( ) baz
+--  ( ) foo
+-- @
+data SearchSelect = SearchSelect
+    { searchWidget :: TextInput
+    -- ^ The field the user uses to search the options.
+    , selectWidget :: Select
+    -- ^ The options presented to a user. Searched through via the searchField.
+    -- This widget should /not/ be used to query all available options;
+    -- it only shows the filtered output.
+    , globalOptions :: Map Text Bool
+    -- ^ Used internally to keep track of all options and their statuses, even
+    -- when not shown to the user.
+    , showOnNoInput :: Bool
+    -- ^ Whether to show the options when the user has yet to type any input.
+    , showCount :: Int
+    -- ^ The maximum number of items to show at any given time.
+    }
+    deriving stock (Generic)
+
+filterSS :: Text -> SearchSelect -> SearchSelect
+filterSS t ss@(SearchSelect{selectWidget, globalOptions, showCount}) =
+    ss{selectWidget = newSelect}
+  where
+    filteredEntries =
+        take showCount
+            . Map.toList
+            $ Map.filterWithKey (const . T.isInfixOf t) globalOptions
+    filterLen = length filteredEntries
+    newCursorRow
+        | selectWidget.cursorRow > filterLen =
+            -- 'max' prevents hookups, since 'Select' is 1-indexed
+            max filterLen 1
+        | otherwise = selectWidget.cursorRow
+    newSelect = selectWidget{options = filteredEntries, cursorRow = newCursorRow}
+
+updateGlobal :: Select -> Map Text Bool -> Map Text Bool
+updateGlobal sel
+    | sel.multiselect = Map.union updates
+    | otherwise = Map.union updates . Map.map (const False)
+  where
+    updates = Map.fromList $ filter snd sel.options
+
+searchVal :: TextInput -> Text
+searchVal = RopeZipper.toText . (.value)
+
+highlightSelected :: Select -> Select
+highlightSelected s@(Select{options, cursorRow}) = s{options = newOptions}
+  where
+    i = max (cursorRow - 1) 0
+    (t, b) = options !! i
+    stylizedSelOpt = ("[" <> t <> "]", b)
+    newOptions
+        | null options = []
+        | otherwise = take i options ++ [stylizedSelOpt] ++ drop (i + 1) options
```

Comment:
Why not `traverse` with `ix`?

Reply:
`traverse` isnt required here (what are we traversing?), but `ix` is. this is fixed.